### ISSUE TITLE
[Bug]: Deadlock while (concurrent) inserts into a hypertable that's linked with a foreign key

### ISSUE THREAD

**Issue by hardikm10** [timescale/timescaledb#4800]
### What type of bug is this?

Performance issue

### What subsystems and features are affected?

Data ingestion

### What happened?

The use-case was to migrating a few million existing database entries from another environment. As the migration tool was made more parallel, we observed deadlocks.

There's a table called "scans" and it has a foreign key reference to another table called "sessions" (which is not a hypertable).

It's hitting deadlock in the PLPGSQL function which inserts data into these tables. Have observed two remedies so far:
1. if we break the FOREIGN KEY relationship (a scan references a session), then everything works fine.
2. if we don't turn the Scans table into a hypertable, everything works fine.

Compression isn't used. They don't have triggers and the insertion logic isn't very complex too.

A brief around the migration tool :
There is a loop which queries the old database for session records. The loop dispatches the sessions using a go channel (a queue, basically) to one of four (configurable) goroutine workers. Each worker picks up a session and processes it complete. So no two workers ever work on the same session. The worker queries the old database for the scans which comprise the session in 2000 scan portions (configurable). It then creates a batch, and writes that batch of scans to Timescale.

[dead_lock_no.log](https://github.com/timescale/timescaledb/files/9723363/dead_lock_no.log)

### TimescaleDB version affected

2.7.0

### PostgreSQL version used

14.1

### What operating system did you use?

MST

### What installation method did you use?

Other

### What platform did you run on?

Managed Service for TimescaleDB (MST/Aiven)

### Relevant log output and stack trace

```bash
2022-07-11 06:05:11.394 UTC [86] ERROR:  deadlock detected
2022-07-11 06:05:11.394 UTC [86] DETAIL:  Process 86 waits for ShareUpdateExclusiveLock on relation 17617 of database 13756; blocked by process 85.
	Process 85 waits for ShareRowExclusiveLock on relation 17599 of database 13756; blocked by process 86.
	Process 86: SELECT scan_uuid::uuid, created_scan::bool, created_session::bool FROM insert_session_legacy_parcel('b5c46bdf-a4fc-42ef-ac42-30d1780df2cf','2021-11-13 01:00:00','0abc3007-23fc-499e-8980-40f04e9b2938','mjipCIznKLuPtfpvGuQz','WixGPbfwAD',NULL,'{"imported": true}');

17617 is the 'scans' hypertable
17599 is the 'sessions' table


	Process 85: SELECT scan_uuid::uuid, created_scan::bool, created_session::bool FROM insert_session_legacy_parcel('b4f0c39c-6b0c-45f8-a21b-b9b783f629a9','2021-06-24 01:00:00','cc46e0ef-3fe3-4bd7-b57c-a427bceb4415','lBzgbaiCMRAjWwhTHctc','uAxhxKQFDa',NULL,'{"imported": true}');
2022-07-11 06:05:11.394 UTC [86] HINT:  See server log for query details.
2022-07-11 06:05:11.394 UTC [86] CONTEXT:  SQL statement "INSERT INTO
	    scans (session_id, scanned, primary_symbol, scan_uuid, scangate, carrier_data, xattrs)
	  VALUES
	    (i_session_id, i_scanned, i_symbol, t_scan_uuid, i_scangate, i_carrier_data, i_xattrs)
	  ON CONFLICT DO NOTHING
	  RETURNING scans.scan_uuid"
	PL/pgSQL function insert_session_legacy_parcel(uuid,timestamp without time zone,uuid,text,text,jsonb,jsonb) line 32 at SQL statement
2022-07-11 06:05:11.394 UTC [86] STATEMENT:  SELECT scan_uuid::uuid, created_scan::bool, created_session::bool FROM insert_session_legacy_parcel('b5c46bdf-a4fc-42ef-ac42-30d1780df2cf','2021-11-13 01:00:00','0abc3007-23fc-499e-8980-40f04e9b2938','mjipCIznKLuPtfpvGuQz','WixGPbfwAD',NULL,'{"imported": true}');

Also,
2022-07-11 06:05:14.726 UTC [86] CONTEXT:  SQL statement " ALTER TABLE _timescaledb_internal._hyper_1_20_chunk ADD CONSTRAINT "20_39_scans_pkey" PRIMARY KEY (session_id, scanned, primary_symbol) "
2022-07-11 06:05:15.721 UTC [85] ERROR:  deadlock detected
2022-07-11 06:05:15.721 UTC [85] DETAIL:  Process 85 waits for ShareUpdateExclusiveLock on relation 17617 of database 13756; blocked by process 86.

```

### How can we reproduce the bug?

```bash
(Reduced) Test case to reproduce :
Two files (about 1000 statements each) that needs to be run concurrently, containing calls to the above PLGPGSQL function with different values.

The reproduction statements could be find in the linked SDC(private) repository due to copyright issues.
```

**Comment by fabriziomello** [comment]
@hardikm10 where is the link for the reproduction statements, I didn't found it on the issue

**Comment by hardikm10** [comment]
@fabriziomello Like we discussed, the statement can be found at https://github.com/timescale/Support-Dev-Collab/issues/495

**Comment by fabriziomello** [comment]
@hardikm10 I've tried with 2.7.2, 2.8.1 and 2.9.0-dev branches and can't reproduce the problem in my local environment. Do u have some environment that I can access where this deadlock happen???

**Comment by hardikm10** [comment]
I'll DM you the $URI value of a service I recreated the deadlock on. 

insert_one.txt and insert_two.txt are already available in the SDC report.

![Image 2022-10-19 at 12 05 30 PM](https://user-images.githubusercontent.com/8678005/196614734-f6ccc1a2-14f8-4dd3-93bf-2fec956bb1ab.jpg)

**Comment by jvanns** [comment]
Hi. I've been asked to contribute my findings to this issue since I appear to have exposed the same bug.

### Our Setup
16 concurrent writers (libpq)
Docker timescaledb:2.8.0-pg14
Slack thread where first discussed; https://timescaledb.slack.com/archives/C63MYDZ35/p1666363249133249?thread_ts=1665650616.295029&cid=C63MYDZ35

'devices' is just a regular postgres table
'metrics' is a hypertable (with space and time partitioning)
There is a foreign key reference from metrics to devices.

We're using libpq COPY statements (in binary mode, if that matters) exclusively and data is only written (it's effectively ingest), never read (explicitly anyway).

### The error
```
ERROR:  deadlock detected
DETAIL:
Process 110 waits for ShareLock on transaction 810; blocked by process 109.
Process 109 waits for ShareUpdateExclusiveLock on relation 18376 of database 17741; blocked by process 108.
Process 108 waits for ShareRowExclusiveLock on relation 17775 of database 17741; blocked by process 110.
Process 110: INSERT INTO devices (ts,id,name) SELECT ts,id,name FROM devices_tt ORDER BY id ON CONFLICT (id) DO UPDATE SET ts = EXCLUDED.ts
Process 109: COPY metrics (ts,device_id,code,result,num,duration,flags) FROM STDIN WITH (FORMAT binary)
Process 108: COPY metrics (ts,device_id,code,result,num,duration,flags) FROM STDIN WITH (FORMAT binary)
```

Digging further reveals that processes 108 & 109, which I think we can conclude are the source of the problem, are in fact running these statements 'under the hood';

```
[108] CONTEXT:  SQL statement "ALTER TABLE _timescaledb_internal._hyper_1_1_chunk ADD CONSTRAINT "1_2_metrics_device_id_fkey" FOREIGN KEY (device_id) REFERENCES public.devices(id) ON DELETE CASCADE DEFERRABLE INITIALLY DEFERRED"
PL/pgSQL function _timescaledb_internal.chunk_constraint_add_table_constraint(_timescaledb_catalog.chunk_constraint) line 58 at EXECUTE COPY metrics, line 1
```

I can reproduce this every single time, totalling between 100-200 errors across 12000 statements, from a stress simulation that bulk-uploads around 12,000,000 tuples in 15 minutes. If I simply comment-out the create_hypertable() and keep metrics as a regular table, the problem completely disappears and I get 0 errors. Every time. I've now run it about 20 times over the past 2 days! Again, there are 16 concurrent sessions to the DB handling this traffic.

I'm afraid this time I am unable to help out further by producing an MRE without our code & data etc. If I find the time later then I shall do so but for now, this is my tuppence!

**Comment by jvanns** [comment]
Has anyone on the TimescaleDB dev team been able to reproduce this? Unsure if my update to the @hardikm10 issue has helped but it definitely does affect us too.

**Comment by konskov** [comment]
Hi @jvanns, thank you for contributing your findings. Indeed it is the same issue as reported above, which we have been able to reproduce. We do not have a fix at the moment, but we are investigating.

**Comment by konskov** [comment]
A quick update on what appears to be the cause of the deadlock here.
Let's take this minimal example as reference:
```
create table customer (
    id int primary key, 
    name text
);

create table card (
    created timestamptz not null,
    customer_id int not null references customer(id),
    primary key(customer_id, created)
);

select create_hypertable('card', 'created');
```
Now let's look at the locks taken:
```
-- session 1
insert into customer (id, name) values
(1, 'konskov');
```
Now S1 is granted a `RowExclusiveLock` on customer. 
```
-- session 2
insert into customer (id, name) values
(2, 'somename');
```
Now S2 is granted a `RowExclusiveLock` on customer. 
```
-- session 1
insert into card(created, customer_id) values (now(), 1);
```
Now S1 is going to create a new chunk for card. 
To do that it is going to request a number of locks.
But the most noteworthy here are: 
* `ShareUpdateExclusiveLock` on card, because we take that lock to serialize chunk creation on the parent table.
* `ShareRowExclusiveLock` on the new chunk of card, because it is required by the `ALTER TABLE ... ADD CONSTRAINT ... FOREIGN KEY` command
* `ShareRowExclusiveLock` on customer, because it is required by the `ALTER TABLE ... ADD CONSTRAINT ... FOREIGN KEY` command (it is also required to lock the referenced table). This conflicts with the `RowExclusiveLock` held by S2 and is not granted. 
```
-- session 2
insert into card(created, customer_id) values (now(), 2);
```
Same as above, the `ShareUpdateExclusiveLock` is requested on card, attempting to create a new chunk, but cannot be obtained as it conflicts with the same lock held by S1.  
And we end up with a deadlock, where S1 waits for `ShareRowExclusiveLock(customer)` and S2 waits for `ShareUpdateExclusiveLock(card)`.

**Comment by jvanns** [comment]
Thanks for the update @konskov - that certainly sounds like a good step-wise diagnosis of the problem to me!

**Comment by jvanns** [comment]
Hello TSDB devs! Just wondering if there is any further news on this issue? Is it being actively investigated/debugged? If you have docker builds with any test code in, I'd be happy to test it - our internal (scale) simulations can trigger this very easily. Cheers.

**Comment by shhnwz** [comment]
Hello @jvanns ,
Thanks for the extensive analysis, we observed that multiple sites reported this issue. For the sake of simplicity we have clubbed them to a single issue. You can track the progress under this link https://github.com/timescale/timescaledb/issues/5342
